lib/omniauth-ethereum.rb

```diff
@@ -13,22 +14,54 @@ class Ethereum
 
       # the `eth_address` will be the _fake_ unique identifier for the Ethereum strategy
       option :uid_field, :eth_address
+      option :fields, [:eth_message, :eth_address, :eth_signature]
+      option :uid_field, :eth_address
 
-      # the omniauth request phase
       def request_phase
-
-        # helper omniauth form to gather required data
-        form = OmniAuth::Form.new :title => "Ethereum Authentication", :url => callback_path
+        form = OmniAuth::Form.new :title => 'Ethereum Authentication', :url => callback_path
         options.fields.each do |field|
 
           # these fields are read-only and will be filled by javascript in the process
-          form.text_field field.to_s.capitalize.gsub("_", " "), field.to_s, readonly: true, class: field.to_s
+          # we were using readonly here before, why?
+          if field == :eth_message
+            form.label_field 'Eth message', 'eth_message'
+            form.html("<input type='hidden' id='eth_message' name='eth_message' value='#{now}' />")
+          else
+            form.text_field field.to_s.capitalize.tr('_', ' '), field.to_s
+          end
+
         end
 
         # the form button will be heavy on javascript, requesting account, nonce, and signature before submission
-        form.button "Sign-In with Ethereum", class: "eth_connect"
+        form.button 'Sign In'
         form.to_response
       end
+
+      def callback_phase
+        address = request.params['eth_address'].downcase
+        message = request.params['eth_message']
+        signature = request.params['eth_signature']
+        signature_pubkey = Eth::Key.personal_recover message, signature
+        signature_address = (Eth::Utils.public_key_to_address signature_pubkey).downcase
+
+        unix_time = message.scan(/\d+/).first.to_i
+        ten_min = 10 * 60
+        return fail!(:invalid_nonce) unless unix_time + ten_min >= now && unix_time - ten_min <= now
```

Comment:
I would rename this to `:invalid_time` because `nonce` is something different.

https://github.com/q9f/ethereum-on-rails/blob/main/app/controllers/sessions_controller.rb#L38-L49

We check three things:
1. request time to make sure the signature did not expire
2. the nonce (must come from the rails backend, we can implement this logic later though)
3. the signature recovering the correct address

For now, this is fine, but we need to make sure to add nonce checks in future.

Reply:
Good point on the naming conflict.

In this case, I've drifted from the original spec a bit and was using time as a nonce instead of something specific to the user. The primary goal of the nonce is to prevent replay attacks right? Are there other goals in attaching this to a user account? I was thinking that we do prevent replay attacks here because an `eth_message` is only valid within 10 minutes of the request and is included in the signature (not just as a form input).

I see a comment [here](https://github.com/q9f/ethereum-on-rails/blob/794ecbc2c0d522dd0a3f51d39d6131b4ce95f29a/app/controllers/sessions_controller.rb#L54) `              # rotate the random nonce to prevent signature spoofing`. I think I'm missing: how would signature spoofing work if we have validated the address through the signature? Wouldn't the integrator of this strategy validate that the address in the strategies `uid`/address is the address they have stored in the user's account?

Separately, I'm thinking there is probably a way to expose a config to include user info, but need to look in to it more. Maybe it would look like [how `omniauth-identity` attaches to ActiveRecord](https://github.com/omniauth/omniauth-identity/blob/master/spec/omniauth/strategies/identity_spec.rb)?

Reply:
> In this case, I've drifted from the original spec a bit and was using time as a nonce instead of something specific to the user.

This is not a problem, time as nonce is "good enough" for now.

> The primary goal of the nonce is to prevent replay attacks right? Are there other goals in attaching this to a user account? I was thinking that we do prevent replay attacks here because an `eth_message` is only valid within 10 minutes of the request and is included in the signature (not just as a form input).

This would allow reusing the signature (assuming an attacker compromised the user) for up to 10 minutes to gain access. The time is only one aspect of protection here. However, the signature remains valid more than once.

> I see a comment [here](https://github.com/q9f/ethereum-on-rails/blob/794ecbc2c0d522dd0a3f51d39d6131b4ce95f29a/app/controllers/sessions_controller.rb#L54) ` # rotate the random nonce to prevent signature spoofing`. I think I'm missing: how would signature spoofing work if we have validated the address through the signature? Wouldn't the integrator of this strategy validate that the address in the strategies `uid`/address is the address they have stored in the user's account?

If you have a valid signature matching an address on our user database, we assume the user is authenticated. always rotating the nonce after each authentication ensures that a signature can only be used once to sign in and is invalid afterward (because the nonce would mismatch on a 2nd attempt - here: maybe an attacker?).

Reply:
Ah got it. That makes sense.